From 98f03d9e33a041ace2d282bdd36b7412365d93ec Mon Sep 17 00:00:00 2001
From: Giuseppe D'Angelo <giuseppe.dangelo@kdab.com>
Date: Sun, 22 Nov 2020 16:02:50 +0100
Subject: [PATCH 097/171] QStackedLayout: fix a memory leak

A layout takes ownership of the QLayoutItem passed to addItem.

Change-Id: Iea3c64453f3cbf4f1b5660e505c1a35dca537833
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit 1fcc882d7ca564f17e49b8d51adc43eb5b97c57b)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/widgets/kernel/qstackedlayout.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/widgets/kernel/qstackedlayout.cpp b/src/widgets/kernel/qstackedlayout.cpp
index f87fe96cc0..2f3d6f39cb 100644
--- a/src/widgets/kernel/qstackedlayout.cpp
+++ b/src/widgets/kernel/qstackedlayout.cpp
@@ -44,6 +44,8 @@
 #include "private/qwidget_p.h"
 #include "private/qlayoutengine_p.h"
 
+#include <memory>
+
 QT_BEGIN_NAMESPACE
 
 class QStackedLayoutPrivate : public QLayoutPrivate
@@ -421,13 +423,13 @@ int QStackedLayout::count() const
 */
 void QStackedLayout::addItem(QLayoutItem *item)
 {
+    std::unique_ptr<QLayoutItem> guard(item);
     QWidget *widget = item->widget();
     if (Q_UNLIKELY(!widget)) {
         qWarning("QStackedLayout::addItem: Only widgets can be added");
         return;
     }
     addWidget(widget);
-    delete item;
 }
 
 /*!
-- 
2.31.1

